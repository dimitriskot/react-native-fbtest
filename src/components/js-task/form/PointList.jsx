import React from 'react';
import { SortableContainer } from 'react-sortable-hoc';
import PointItem from "./PointItem";

const PointList = (props) => {
	const { points, onDeleteButtonClick } = props;
	return (
		<ul className="map-form__list outline-1" >
			{points.map((point, i) => (
				<PointItem key={point.id} point={point} index={i} onDeleteButtonClick={onDeleteButtonClick.bind(this, point.id)} />
			))}
		</ul>
	)
};

export default SortableContainer(PointList);