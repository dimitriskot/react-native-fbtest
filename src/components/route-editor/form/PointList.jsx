import React from "react";
import PropTypes from "prop-types";
import { SortableContainer } from "react-sortable-hoc";
import PointItem from "./PointItem";

const PointList = (props) => {
  const { points, onDeleteButtonClick } = props;
  return (
    <ul className={"map-form__list"}>
      {points.map((point, i) => (
        <PointItem
          key={point.id}
          point={point}
          index={i}
          onDeleteButtonClick={onDeleteButtonClick.bind(this, point.id)}
        />
      ))}
    </ul>
  );
};

PointList.propTypes = {
  points: PropTypes.array,
  onDeleteButtonClick: PropTypes.func
};

export default SortableContainer(PointList);
