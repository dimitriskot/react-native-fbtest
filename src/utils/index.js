export const numberDeclension = (number, titles) => {
  return titles[
    number % 10 === 1 && number % 100 !== 11
      ? 0
      : number % 10 >= 2 && number % 10 <= 4 && (number % 100 < 10 || number % 100 >= 20)
        ? 1
        : 2
  ];
};

export const randomId = () => {
  return `_${Math.random()
    .toString(36)
    .substr(2, 9)}`;
};
