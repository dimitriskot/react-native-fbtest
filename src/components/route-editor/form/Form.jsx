import React from "react";
import PropTypes from "prop-types";
import PointList from "./PointList";

const MapFormComponent = (props) => {
  const {
    points,
    value,
    handleValueChange,
    handleEnterPress,
    handleCreateButtonClick,
    handleDeleteButtonClick,
    handleSortEnd
  } = props;

  return (
    <div className={"map-form"}>
      <div className={"map-form__header"}>
        <p className={"map-form__header-text"}>Создать точку маршрута:</p>
        <input
          className={"map-form__header-input"}
          type={"text"}
          name={"point"}
          value={value}
          onChange={handleValueChange}
          onKeyDown={handleEnterPress}
          maxLength={50}
          autoComplete={"off"}
          required
        />
        <button
          className={"map-form__header-button"}
          onClick={handleCreateButtonClick}
          disabled={!value}
        >
            Создать
        </button>
      </div>
      <PointList
        points={points}
        onDeleteButtonClick={handleDeleteButtonClick}
        onSortEnd={handleSortEnd}
      />
    </div>
  );

};

MapFormComponent.propTypes = {
  points: PropTypes.array.isRequired,
  value: PropTypes.string.isRequired,
  handleValueChange: PropTypes.func.isRequired,
  handleEnterPress: PropTypes.func.isRequired,
  handleCreateButtonClick: PropTypes.func.isRequired,
  handleDeleteButtonClick: PropTypes.func.isRequired,
  handleSortEnd: PropTypes.func.isRequired
};

export default MapFormComponent;
