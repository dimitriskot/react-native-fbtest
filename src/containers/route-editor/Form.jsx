import React from "react";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import { arrayMove } from "react-sortable-hoc";
import routeEditorActions from "../../store/actions/route-editor";
import { randomId } from "../../utils";
import store from "../../store";
import MapFormComponent from "../../components/route-editor/form/Form";

class MapFormContainer extends React.Component {

  ENTER_KEY = "Enter";

  state = { value: "" }

  handleValueChange = (e) => {
    const value = e.currentTarget.value;
    this.setState({ value });
  };

  handleCreateButtonClick = () => {
    const { addPoint } = this.props;
    const point = this.generatePoint();
    addPoint({ point });
    const value = "";
    this.setState({ value });
    this.getRoute();
  };

  handleEnterPress = (e) => {
    const { value } = this.state;
    if (e.key !== this.ENTER_KEY || !value) {
      return;
    }
    this.handleCreateButtonClick();
  };

  handleDeleteButtonClick = (id) => {
    const { deletePoint } = this.props;
    deletePoint({ id });
    this.getRoute();
  };

  handleSortEnd = ({ oldIndex, newIndex }) => {
    const { map: { points }, changePointsOrder } = this.props;
    const newPoints = arrayMove(points, oldIndex, newIndex);
    changePointsOrder({ newPoints });
    this.getRoute();
  };

  generatePoint = () => {
    const { value } = this.state;
    const { map: { center } } = this.props;
    return {
      id: randomId(),
      title: value,
      position: center,
      draggable: true,
      handleDragEnd: (index, marker) => this.handleMarkerDragEnd(index, marker)
    };
  };

  handleMarkerDragEnd = (id, { latLng }) => {
    const { changePointPosition } = this.props;
    changePointPosition({ id, latLng });
    this.getRoute();
  };

  getRoute = () => {
    const { getDirections } = this.props;
    const { routeEditor: { points } } = store.getState();
    const isPoints = points.length > 0;
    if (isPoints) {
      const google = window.google;
      const DirectionsService = new google.maps.DirectionsService();
      const waypoints = points.map(({ position }) => ({
        location: position,
        stopover: true
      }));

      const origin = points[0].position;
      const destination = points[points.length - 1].position;

      DirectionsService.route(
        {
          origin,
          destination,
          waypoints,
          optimizeWaypoints: false,
          travelMode: google.maps.TravelMode.WALKING
        },
        (directions, status) => {
          if (status === google.maps.DirectionsStatus.OK) {
            getDirections({ directions });
          } else {
            console.error(`Directions request failed due to ${directions}`);
          }
        },
      );
    }
  };

  render() {
    const { value } = this.state;
    const { map: { points } } = this.props;

    return (
      <MapFormComponent
        value={value}
        points={points}
        handleValueChange={this.handleValueChange}
        handleEnterPress={this.handleEnterPress}
        handleCreateButtonClick={this.handleCreateButtonClick}
        handleDeleteButtonClick={this.handleDeleteButtonClick}
        handleSortEnd={this.handleSortEnd}
      />
    );
  }

}

MapFormContainer.propTypes = {
  map: PropTypes.shape({
    center: PropTypes.object.isRequired,
    points: PropTypes.array.isRequired
  }).isRequired,
  addPoint: PropTypes.func.isRequired,
  deletePoint: PropTypes.func.isRequired,
  getDirections: PropTypes.func.isRequired,
  changePointPosition: PropTypes.func.isRequired,
  changePointsOrder: PropTypes.func.isRequired
};

const mapStateToProps = (state) => {
  return { map: state.routeEditor };
};

const MapForm = connect(
  mapStateToProps,
  routeEditorActions
)(MapFormContainer);

export default MapForm;
