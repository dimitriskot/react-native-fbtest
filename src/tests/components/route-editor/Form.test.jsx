import React from "react";
import { shallow } from "enzyme";
import Form from "../../../components/route-editor/form/Form";
import PointList from "../../../components/route-editor/form/PointList";
import { randomId } from "../../../utils";

describe("Form component", () => {

  const id = randomId();
  const props = {
    points: [{ id }],
    value: "point",
    handleValueChange: jest.fn(),
    handleEnterPress: jest.fn(),
    handleCreateButtonClick: jest.fn(),
    handleDeleteButtonClick: jest.fn(),
    handleSortEnd: jest.fn()
  };
  const wrapper = shallow(<Form {...props} />);

  it("render Form", () => {
    expect(wrapper.render()).toMatchSnapshot();
    expect(wrapper.find(PointList)).toHaveLength(1);
    expect(wrapper.find("input.map-form__header-input")).toHaveLength(1);
  });

  it("handleValueChange", () => {
    const input = wrapper.find("input.map-form__header-input");
    input.simulate("change", { target: { value: props.value } });
    expect(props.handleValueChange).toBeCalledTimes(1);
    expect(props.handleValueChange).toHaveBeenCalledWith({ target: { value: props.value } });
  });

  it("handleEnterPress", () => {
    const ENTER_KEY = "Enter";
    const input = wrapper.find("input.map-form__header-input");
    input.simulate("keyDown", { key: ENTER_KEY });
    expect(props.handleEnterPress).toBeCalledTimes(1);
    expect(props.handleEnterPress).toHaveBeenCalledWith({ key: ENTER_KEY });
  });

});
