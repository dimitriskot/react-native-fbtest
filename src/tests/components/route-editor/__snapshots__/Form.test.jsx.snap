// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form component render Form 1`] = `
initialize {
  "0": Object {
    "attribs": Object {
      "class": "map-form",
    },
    "children": Array [
      Object {
        "attribs": Object {
          "class": "map-form__header",
        },
        "children": Array [
          Object {
            "attribs": Object {
              "class": "map-form__header-text",
            },
            "children": Array [
              Object {
                "data": "Создать точку маршрута:",
                "next": null,
                "parent": [Circular],
                "prev": null,
                "type": "text",
              },
            ],
            "name": "p",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": Object {
              "attribs": Object {
                "autocomplete": "off",
                "class": "map-form__header-input",
                "maxlength": "50",
                "name": "point",
                "required": "",
                "type": "text",
                "value": "point",
              },
              "children": Array [],
              "name": "input",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": Object {
                "attribs": Object {
                  "class": "map-form__header-button",
                },
                "children": Array [
                  Object {
                    "data": "Создать",
                    "next": null,
                    "parent": [Circular],
                    "prev": null,
                    "type": "text",
                  },
                ],
                "name": "button",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": null,
                "parent": [Circular],
                "prev": [Circular],
                "type": "tag",
                "x-attribsNamespace": Object {
                  "class": undefined,
                },
                "x-attribsPrefix": Object {
                  "class": undefined,
                },
              },
              "parent": [Circular],
              "prev": [Circular],
              "type": "tag",
              "x-attribsNamespace": Object {
                "autocomplete": undefined,
                "class": undefined,
                "maxlength": undefined,
                "name": undefined,
                "required": undefined,
                "type": undefined,
                "value": undefined,
              },
              "x-attribsPrefix": Object {
                "autocomplete": undefined,
                "class": undefined,
                "maxlength": undefined,
                "name": undefined,
                "required": undefined,
                "type": undefined,
                "value": undefined,
              },
            },
            "parent": [Circular],
            "prev": null,
            "type": "tag",
            "x-attribsNamespace": Object {
              "class": undefined,
            },
            "x-attribsPrefix": Object {
              "class": undefined,
            },
          },
          Object {
            "attribs": Object {
              "autocomplete": "off",
              "class": "map-form__header-input",
              "maxlength": "50",
              "name": "point",
              "required": "",
              "type": "text",
              "value": "point",
            },
            "children": Array [],
            "name": "input",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": Object {
              "attribs": Object {
                "class": "map-form__header-button",
              },
              "children": Array [
                Object {
                  "data": "Создать",
                  "next": null,
                  "parent": [Circular],
                  "prev": null,
                  "type": "text",
                },
              ],
              "name": "button",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": null,
              "parent": [Circular],
              "prev": [Circular],
              "type": "tag",
              "x-attribsNamespace": Object {
                "class": undefined,
              },
              "x-attribsPrefix": Object {
                "class": undefined,
              },
            },
            "parent": [Circular],
            "prev": Object {
              "attribs": Object {
                "class": "map-form__header-text",
              },
              "children": Array [
                Object {
                  "data": "Создать точку маршрута:",
                  "next": null,
                  "parent": [Circular],
                  "prev": null,
                  "type": "text",
                },
              ],
              "name": "p",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": [Circular],
              "parent": [Circular],
              "prev": null,
              "type": "tag",
              "x-attribsNamespace": Object {
                "class": undefined,
              },
              "x-attribsPrefix": Object {
                "class": undefined,
              },
            },
            "type": "tag",
            "x-attribsNamespace": Object {
              "autocomplete": undefined,
              "class": undefined,
              "maxlength": undefined,
              "name": undefined,
              "required": undefined,
              "type": undefined,
              "value": undefined,
            },
            "x-attribsPrefix": Object {
              "autocomplete": undefined,
              "class": undefined,
              "maxlength": undefined,
              "name": undefined,
              "required": undefined,
              "type": undefined,
              "value": undefined,
            },
          },
          Object {
            "attribs": Object {
              "class": "map-form__header-button",
            },
            "children": Array [
              Object {
                "data": "Создать",
                "next": null,
                "parent": [Circular],
                "prev": null,
                "type": "text",
              },
            ],
            "name": "button",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": null,
            "parent": [Circular],
            "prev": Object {
              "attribs": Object {
                "autocomplete": "off",
                "class": "map-form__header-input",
                "maxlength": "50",
                "name": "point",
                "required": "",
                "type": "text",
                "value": "point",
              },
              "children": Array [],
              "name": "input",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": [Circular],
              "parent": [Circular],
              "prev": Object {
                "attribs": Object {
                  "class": "map-form__header-text",
                },
                "children": Array [
                  Object {
                    "data": "Создать точку маршрута:",
                    "next": null,
                    "parent": [Circular],
                    "prev": null,
                    "type": "text",
                  },
                ],
                "name": "p",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": [Circular],
                "parent": [Circular],
                "prev": null,
                "type": "tag",
                "x-attribsNamespace": Object {
                  "class": undefined,
                },
                "x-attribsPrefix": Object {
                  "class": undefined,
                },
              },
              "type": "tag",
              "x-attribsNamespace": Object {
                "autocomplete": undefined,
                "class": undefined,
                "maxlength": undefined,
                "name": undefined,
                "required": undefined,
                "type": undefined,
                "value": undefined,
              },
              "x-attribsPrefix": Object {
                "autocomplete": undefined,
                "class": undefined,
                "maxlength": undefined,
                "name": undefined,
                "required": undefined,
                "type": undefined,
                "value": undefined,
              },
            },
            "type": "tag",
            "x-attribsNamespace": Object {
              "class": undefined,
            },
            "x-attribsPrefix": Object {
              "class": undefined,
            },
          },
        ],
        "name": "div",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": Object {
          "attribs": Object {
            "class": "map-form__list",
          },
          "children": Array [
            Object {
              "attribs": Object {
                "class": "map-form__list-item",
              },
              "children": Array [
                Object {
                  "attribs": Object {
                    "class": "map-form__list-item__text",
                  },
                  "children": Array [],
                  "name": "p",
                  "namespace": "http://www.w3.org/1999/xhtml",
                  "next": Object {
                    "attribs": Object {
                      "class": "map-form__list-item__button",
                    },
                    "children": Array [],
                    "name": "button",
                    "namespace": "http://www.w3.org/1999/xhtml",
                    "next": null,
                    "parent": [Circular],
                    "prev": [Circular],
                    "type": "tag",
                    "x-attribsNamespace": Object {
                      "class": undefined,
                    },
                    "x-attribsPrefix": Object {
                      "class": undefined,
                    },
                  },
                  "parent": [Circular],
                  "prev": null,
                  "type": "tag",
                  "x-attribsNamespace": Object {
                    "class": undefined,
                  },
                  "x-attribsPrefix": Object {
                    "class": undefined,
                  },
                },
                Object {
                  "attribs": Object {
                    "class": "map-form__list-item__button",
                  },
                  "children": Array [],
                  "name": "button",
                  "namespace": "http://www.w3.org/1999/xhtml",
                  "next": null,
                  "parent": [Circular],
                  "prev": Object {
                    "attribs": Object {
                      "class": "map-form__list-item__text",
                    },
                    "children": Array [],
                    "name": "p",
                    "namespace": "http://www.w3.org/1999/xhtml",
                    "next": [Circular],
                    "parent": [Circular],
                    "prev": null,
                    "type": "tag",
                    "x-attribsNamespace": Object {
                      "class": undefined,
                    },
                    "x-attribsPrefix": Object {
                      "class": undefined,
                    },
                  },
                  "type": "tag",
                  "x-attribsNamespace": Object {
                    "class": undefined,
                  },
                  "x-attribsPrefix": Object {
                    "class": undefined,
                  },
                },
              ],
              "name": "li",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": null,
              "parent": [Circular],
              "prev": null,
              "type": "tag",
              "x-attribsNamespace": Object {
                "class": undefined,
              },
              "x-attribsPrefix": Object {
                "class": undefined,
              },
            },
          ],
          "name": "ul",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": null,
          "parent": [Circular],
          "prev": [Circular],
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
          },
        },
        "parent": [Circular],
        "prev": null,
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
        },
      },
      Object {
        "attribs": Object {
          "class": "map-form__list",
        },
        "children": Array [
          Object {
            "attribs": Object {
              "class": "map-form__list-item",
            },
            "children": Array [
              Object {
                "attribs": Object {
                  "class": "map-form__list-item__text",
                },
                "children": Array [],
                "name": "p",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": Object {
                  "attribs": Object {
                    "class": "map-form__list-item__button",
                  },
                  "children": Array [],
                  "name": "button",
                  "namespace": "http://www.w3.org/1999/xhtml",
                  "next": null,
                  "parent": [Circular],
                  "prev": [Circular],
                  "type": "tag",
                  "x-attribsNamespace": Object {
                    "class": undefined,
                  },
                  "x-attribsPrefix": Object {
                    "class": undefined,
                  },
                },
                "parent": [Circular],
                "prev": null,
                "type": "tag",
                "x-attribsNamespace": Object {
                  "class": undefined,
                },
                "x-attribsPrefix": Object {
                  "class": undefined,
                },
              },
              Object {
                "attribs": Object {
                  "class": "map-form__list-item__button",
                },
                "children": Array [],
                "name": "button",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": null,
                "parent": [Circular],
                "prev": Object {
                  "attribs": Object {
                    "class": "map-form__list-item__text",
                  },
                  "children": Array [],
                  "name": "p",
                  "namespace": "http://www.w3.org/1999/xhtml",
                  "next": [Circular],
                  "parent": [Circular],
                  "prev": null,
                  "type": "tag",
                  "x-attribsNamespace": Object {
                    "class": undefined,
                  },
                  "x-attribsPrefix": Object {
                    "class": undefined,
                  },
                },
                "type": "tag",
                "x-attribsNamespace": Object {
                  "class": undefined,
                },
                "x-attribsPrefix": Object {
                  "class": undefined,
                },
              },
            ],
            "name": "li",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": null,
            "parent": [Circular],
            "prev": null,
            "type": "tag",
            "x-attribsNamespace": Object {
              "class": undefined,
            },
            "x-attribsPrefix": Object {
              "class": undefined,
            },
          },
        ],
        "name": "ul",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": null,
        "parent": [Circular],
        "prev": Object {
          "attribs": Object {
            "class": "map-form__header",
          },
          "children": Array [
            Object {
              "attribs": Object {
                "class": "map-form__header-text",
              },
              "children": Array [
                Object {
                  "data": "Создать точку маршрута:",
                  "next": null,
                  "parent": [Circular],
                  "prev": null,
                  "type": "text",
                },
              ],
              "name": "p",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": Object {
                "attribs": Object {
                  "autocomplete": "off",
                  "class": "map-form__header-input",
                  "maxlength": "50",
                  "name": "point",
                  "required": "",
                  "type": "text",
                  "value": "point",
                },
                "children": Array [],
                "name": "input",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": Object {
                  "attribs": Object {
                    "class": "map-form__header-button",
                  },
                  "children": Array [
                    Object {
                      "data": "Создать",
                      "next": null,
                      "parent": [Circular],
                      "prev": null,
                      "type": "text",
                    },
                  ],
                  "name": "button",
                  "namespace": "http://www.w3.org/1999/xhtml",
                  "next": null,
                  "parent": [Circular],
                  "prev": [Circular],
                  "type": "tag",
                  "x-attribsNamespace": Object {
                    "class": undefined,
                  },
                  "x-attribsPrefix": Object {
                    "class": undefined,
                  },
                },
                "parent": [Circular],
                "prev": [Circular],
                "type": "tag",
                "x-attribsNamespace": Object {
                  "autocomplete": undefined,
                  "class": undefined,
                  "maxlength": undefined,
                  "name": undefined,
                  "required": undefined,
                  "type": undefined,
                  "value": undefined,
                },
                "x-attribsPrefix": Object {
                  "autocomplete": undefined,
                  "class": undefined,
                  "maxlength": undefined,
                  "name": undefined,
                  "required": undefined,
                  "type": undefined,
                  "value": undefined,
                },
              },
              "parent": [Circular],
              "prev": null,
              "type": "tag",
              "x-attribsNamespace": Object {
                "class": undefined,
              },
              "x-attribsPrefix": Object {
                "class": undefined,
              },
            },
            Object {
              "attribs": Object {
                "autocomplete": "off",
                "class": "map-form__header-input",
                "maxlength": "50",
                "name": "point",
                "required": "",
                "type": "text",
                "value": "point",
              },
              "children": Array [],
              "name": "input",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": Object {
                "attribs": Object {
                  "class": "map-form__header-button",
                },
                "children": Array [
                  Object {
                    "data": "Создать",
                    "next": null,
                    "parent": [Circular],
                    "prev": null,
                    "type": "text",
                  },
                ],
                "name": "button",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": null,
                "parent": [Circular],
                "prev": [Circular],
                "type": "tag",
                "x-attribsNamespace": Object {
                  "class": undefined,
                },
                "x-attribsPrefix": Object {
                  "class": undefined,
                },
              },
              "parent": [Circular],
              "prev": Object {
                "attribs": Object {
                  "class": "map-form__header-text",
                },
                "children": Array [
                  Object {
                    "data": "Создать точку маршрута:",
                    "next": null,
                    "parent": [Circular],
                    "prev": null,
                    "type": "text",
                  },
                ],
                "name": "p",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": [Circular],
                "parent": [Circular],
                "prev": null,
                "type": "tag",
                "x-attribsNamespace": Object {
                  "class": undefined,
                },
                "x-attribsPrefix": Object {
                  "class": undefined,
                },
              },
              "type": "tag",
              "x-attribsNamespace": Object {
                "autocomplete": undefined,
                "class": undefined,
                "maxlength": undefined,
                "name": undefined,
                "required": undefined,
                "type": undefined,
                "value": undefined,
              },
              "x-attribsPrefix": Object {
                "autocomplete": undefined,
                "class": undefined,
                "maxlength": undefined,
                "name": undefined,
                "required": undefined,
                "type": undefined,
                "value": undefined,
              },
            },
            Object {
              "attribs": Object {
                "class": "map-form__header-button",
              },
              "children": Array [
                Object {
                  "data": "Создать",
                  "next": null,
                  "parent": [Circular],
                  "prev": null,
                  "type": "text",
                },
              ],
              "name": "button",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": null,
              "parent": [Circular],
              "prev": Object {
                "attribs": Object {
                  "autocomplete": "off",
                  "class": "map-form__header-input",
                  "maxlength": "50",
                  "name": "point",
                  "required": "",
                  "type": "text",
                  "value": "point",
                },
                "children": Array [],
                "name": "input",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": [Circular],
                "parent": [Circular],
                "prev": Object {
                  "attribs": Object {
                    "class": "map-form__header-text",
                  },
                  "children": Array [
                    Object {
                      "data": "Создать точку маршрута:",
                      "next": null,
                      "parent": [Circular],
                      "prev": null,
                      "type": "text",
                    },
                  ],
                  "name": "p",
                  "namespace": "http://www.w3.org/1999/xhtml",
                  "next": [Circular],
                  "parent": [Circular],
                  "prev": null,
                  "type": "tag",
                  "x-attribsNamespace": Object {
                    "class": undefined,
                  },
                  "x-attribsPrefix": Object {
                    "class": undefined,
                  },
                },
                "type": "tag",
                "x-attribsNamespace": Object {
                  "autocomplete": undefined,
                  "class": undefined,
                  "maxlength": undefined,
                  "name": undefined,
                  "required": undefined,
                  "type": undefined,
                  "value": undefined,
                },
                "x-attribsPrefix": Object {
                  "autocomplete": undefined,
                  "class": undefined,
                  "maxlength": undefined,
                  "name": undefined,
                  "required": undefined,
                  "type": undefined,
                  "value": undefined,
                },
              },
              "type": "tag",
              "x-attribsNamespace": Object {
                "class": undefined,
              },
              "x-attribsPrefix": Object {
                "class": undefined,
              },
            },
          ],
          "name": "div",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": [Circular],
          "parent": [Circular],
          "prev": null,
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
          },
        },
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
        },
      },
    ],
    "name": "div",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": null,
    "prev": null,
    "root": Object {
      "attribs": Object {},
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-attribsNamespace": Object {},
      "x-attribsPrefix": Object {},
    },
    "type": "tag",
    "x-attribsNamespace": Object {
      "class": undefined,
    },
    "x-attribsPrefix": Object {
      "class": undefined,
    },
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xml": false,
  },
}
`;
